### PR TITLE
sessions until state metric

### PR DESCRIPTION
**Pull request by rudigiesler**

### CODE REVIEW

**Comment by rudigiesler** [comment]
@hodgestar Ready for review.

File: lib/metrics-helper.js [hunk 1]
```diff
@@ -96,6 +110,65 @@ var MetricsHelper = Eventable.extend(function(self, im) {
...
+        self.im.on('session:new', function(e) {
+            /* Increment sessions counter */
+            var metadata = e.im.user.metadata;
+
+            if (typeof metadata[metadata_label] === 'undefined'){
+                metadata[metadata_label] = 0;
+            }
+
+            e.im.user.metadata[metadata_label] += 1;
+            return e.im.user.save();
```
Comment (hodgestar) [inline lib/metrics-helper.js]: The user is saved when the sandbox terminates, so we can perhaps skip saving it here?

**Comment by hodgestar** [comment]
Other than the question about saving, looks good.

I am still worried that we're heading down a route that gives us metric setup that looks like:

``` javascript
mh
  .add.sessions_until_state({state: 'states:test2', action: 'enter'}, 'sessions_until')
  .add.time_until_state({state: 'states:test2', action: 'enter'}, 'time_until')
  .add.avg_session_until_state(... same state ...)
  ...
```

where I still think it'd be useful to have a function like:

``` javascript
mh
  .add.tracker('survey', {
    enter: 'states:start',  // start trigger definition
  }, {
    enter: 'states:test2',  // end trigger definition
  }, {
    sessions_until: 'sessions_until',  // maps metric to label, or null to disable
    time_until: 'time_until',
  })
```

which allows us to put a lot of functionality into a common framework and doesn't require people to respecify start and end events for every metric and remember the names of an ever-growing list of functions?

Thoughts? If we rework things like that, I'm happy to leave it for a separate PR though.

**Comment by rudigiesler** [comment]
With regards to the second pattern, I think we can easily add that pattern. All we do is create an `.add.tracker` function that calls the required metricHelper functions, it shouldn't be too much work to add that pattern. 

I think it might be best to have both patterns available, as the second pattern doesn't allow quick setup using defaults. It also won't support metrics with no state trigger (e.g. `total_unique_users`) or only one state trigger (e.g. `total_state_actions`)

**Comment by hodgestar** [comment]
We could do something similar for things that trigger off a single event
for consistency, e.g. `mh.trigger({ ... event ...}, { ... metrics ... })`?

**Comment by rudigiesler** [comment]
Yes we could do something like that. What do you think of the idea of creating those functions, and just having them call the existing functions?

**Comment by hodgestar** [comment]
Sounds good. That'll allow us to see how the API feels.

**Comment by rudigiesler** [comment]
@hodgestar Created an issue for what was discussed here: https://github.com/praekelt/go-jsbox-metrics-helper/issues/11

**Comment by rudigiesler** [comment]
@hodgestar ready for a review again

**Comment by hodgestar** [comment]
:+1:

**Pull request merged**